File: tests/unit/orgs.test.ts
/**
 * Organization membership resolution (schema v4).
 *
 * Pure input → output: no git, no filesystem, no clock. Everything here is about the union of
 * the two membership directions, the two dangling-reference warnings, and the ordering that
 * makes forge.json byte-identical between runs and between machines.
 */
import os from 'node:os';
import { describe, expect, it } from 'vitest';
import { resolveConfig, type FrznforgeConfigInput } from '../../src/lib/config/index';
import { Organization, emptyForgeData } from '../../src/lib/data/schema';
import { resolveOrganizations, type OrgRepoInput } from '../../src/lib/ingest/orgs';
import { unmatchedOrgContent } from '../../src/lib/routes';

/** A resolved config with just the parts organization resolution reads. */
function configWith(organizations: FrznforgeConfigInput['organizations']) {
  return resolveConfig({ owner: { name: 'Tester', handle: 'tester' }, organizations }, os.tmpdir());
}

/** Shorthand for the scanned-repo pairs `ingest()` hands to `resolveOrganizations`. */
function repo(slug: string, org: string | null = null): OrgRepoInput {
  return { slug, org };
}

describe('resolveOrganizations', () => {
  it('returns nothing when no organizations are configured, whatever the repos say', () => {
    const { organizations, warnings } = resolveOrganizations(configWith([]), [repo('alpha'), repo('beta')]);
    expect(organizations).toEqual([]);
    expect(warnings).toEqual([]);
  });

  it('takes membership from the organization\'s own repos list', () => {
    const cfg = configWith([{ slug: 'acme', name: 'Acme', description: 'Makers of things', repos: ['alpha', 'beta'] }]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('alpha'), repo('beta'), repo('gamma')]);
    expect(warnings).toEqual([]);
    expect(organizations).toEqual([
      { slug: 'acme', name: 'Acme', description: 'Makers of things', repos: ['alpha', 'beta'] },
    ]);
  });

  it('takes membership from a repo source declaring `org`', () => {
    const cfg = configWith([{ slug: 'acme', name: 'Acme' }]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('alpha', 'acme'), repo('beta')]);
    expect(warnings).toEqual([]);
    expect(organizations).toEqual([{ slug: 'acme', name: 'Acme', description: null, repos: ['alpha'] }]);
  });

  it('unions both directions and never lists a repo twice', () => {
    // `alpha` is claimed by the org *and* claims the org — the shipped frznforge.config.ts
    // does exactly this, so the de-duplication path is exercised by the real site too.
    const cfg = configWith([{ slug: 'acme', name: 'Acme', repos: ['alpha', 'beta'] }]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('alpha', 'acme'), repo('beta'), repo('gamma', 'acme')]);
    expect(warnings).toEqual([]);
    expect(organizations[0]!.repos).toEqual(['alpha', 'beta', 'gamma']);
  });

  it('emits an organization with no members rather than dropping it', () => {
    const cfg = configWith([{ slug: 'empty-org', name: 'Empty Org' }]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('alpha')]);
    expect(organizations).toEqual([{ slug: 'empty-org', name: 'Empty Org', description: null, repos: [] }]);
    expect(warnings).toEqual([]);
  });

  it('sorts organizations by slug and members by code point, independent of input order', () => {
    const cfg = configWith([
      { slug: 'zulu', name: 'Zulu', repos: ['ab', 'a-b', 'a2'] },
      { slug: 'alpha-co', name: 'Alpha Co', repos: ['a2'] },
    ]);
    const repos = [repo('a2'), repo('ab'), repo('a-b'), repo('m1', 'alpha-co')];
    const { organizations } = resolveOrganizations(cfg, repos);
    expect(organizations.map((o) => o.slug)).toEqual(['alpha-co', 'zulu']);
    // '-' (0x2D) sorts before digits and letters; `localeCompare` would put 'a-b' last.
    expect(organizations[1]!.repos).toEqual(['a-b', 'a2', 'ab']);
    expect(organizations[0]!.repos).toEqual(['a2', 'm1']);

    // Same inputs in a different order → identical output (determinism).
    const shuffled = resolveOrganizations(cfg, [...repos].reverse());
    expect(shuffled.organizations).toEqual(organizations);
  });

  it('warns `org-unknown-repo` once per dangling entry, in config order, with repo: null', () => {
    const cfg = configWith([
      { slug: 'zulu', name: 'Zulu', repos: ['ghost'] },
      { slug: 'acme', name: 'Acme', repos: ['alpha', 'nope', 'nope'] },
    ]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('alpha')]);
    expect(warnings.map((w) => ({ code: w.code, repo: w.repo }))).toEqual([
      { code: 'org-unknown-repo', repo: null },
      { code: 'org-unknown-repo', repo: null },
    ]);
    expect(warnings[0]!.message).toContain("organization 'zulu' lists repo 'ghost'");
    expect(warnings[1]!.message).toContain("organization 'acme' lists repo 'nope'");
    // the dangling names never reach the artifact
    expect(organizations.map((o) => o.repos)).toEqual([['alpha'], []]);
  });

  it('warns `repo-unknown-org` in slug order, stamped with the repo slug', () => {
    const cfg = configWith([{ slug: 'acme', name: 'Acme' }]);
    const { organizations, warnings } = resolveOrganizations(cfg, [
      repo('zeta', 'typo-org'),
      repo('alpha', 'other-typo'),
      repo('beta', 'acme'),
    ]);
    expect(warnings.map((w) => ({ code: w.code, repo: w.repo }))).toEqual([
      { code: 'repo-unknown-org', repo: 'alpha' },
      { code: 'repo-unknown-org', repo: 'zeta' },
    ]);
    expect(warnings[0]!.message).toContain("repo 'alpha' declares org 'other-typo'");
    expect(organizations[0]!.repos).toEqual(['beta']);
  });

  it('orders warnings organizations-first, then repos', () => {
    const cfg = configWith([{ slug: 'acme', name: 'Acme', repos: ['ghost'] }]);
    const { warnings } = resolveOrganizations(cfg, [repo('alpha', 'nowhere')]);
    expect(warnings.map((w) => w.code)).toEqual(['org-unknown-repo', 'repo-unknown-org']);
  });

  it('lets several organizations claim the same repo', () => {
    // A repo source's `org` names exactly one org, but `organizations[].repos` is a claim and
    // not an exclusive one — so a shared library can legitimately appear under two groups.
    const cfg = configWith([
      { slug: 'acme', name: 'Acme', repos: ['shared'] },
      { slug: 'beta-co', name: 'Beta Co', repos: ['shared'] },
    ]);
    const { organizations, warnings } = resolveOrganizations(cfg, [repo('shared', 'acme')]);
    expect(warnings).toEqual([]);
    expect(organizations.map((o) => o.repos)).toEqual([['shared'], ['shared']]);
  });

  it('merges a duplicated organization slug instead of failing or losing members', () => {
    const cfg = configWith([
      { slug: 'acme', name: 'Acme', description: 'first', repos: ['alpha'] },
      { slug: 'acme', name: 'Acme Again', description: 'second', repos: ['beta'] },
    ]);
    const { organizations } = resolveOrganizations(cfg, [repo('alpha'), repo('beta')]);
    expect(organizations).toEqual([
      { slug: 'acme', name: 'Acme', description: 'first', repos: ['alpha', 'beta'] },
    ]);
  });

  it('produces organizations that validate against the artifact schema', () => {
    const cfg = configWith([{ slug: 'acme', name: 'Acme', repos: ['alpha', 'ghost'] }]);
    const { organizations } = resolveOrganizations(cfg, [repo('alpha'), repo('beta', 'acme')]);
    for (const org of organizations) expect(() => Organization.parse(org)).not.toThrow();
  });
});

describe('unmatchedOrgContent', () => {
  /** An artifact holding just the organizations named. */
  const withOrgs = (...slugs: string[]) => ({
    ...emptyForgeData(),
    organizations: slugs.map((slug) => ({ slug, name: slug, description: null, repos: [] })),
  });

  it('names every orgs markdown file no configured organization claims', () => {
    // One typo in the filename otherwise discards the file's prose, sites, links and pinned
    // repos with no page and no other symptom.
    const data = withOrgs('canadian-coding', 'lonely');
    expect(unmatchedOrgContent(data, ['canadian-codng', 'canadian-coding', 'ghost'])).toEqual([
      'canadian-codng',
      'ghost',
    ]);
  });

  it('is quiet when every file matches, and when there are no files at all', () => {
    const data = withOrgs('canadian-coding', 'lonely');
    expect(unmatchedOrgContent(data, ['lonely', 'canadian-coding'])).toEqual([]);
    expect(unmatchedOrgContent(data, [])).toEqual([]);
    // no organizations configured: every file is unmatched, and still sorted
    expect(unmatchedOrgContent(emptyForgeData(), ['b', 'a'])).toEqual(['a', 'b']);
  });
});

File: internal/ingest/orgs_test.go
package ingest

import (
	"reflect"
	"testing"

	"frznforge/internal/model"
)

func orgSlugs(orgs []model.Organization) []string {
	out := make([]string, len(orgs))
	for i, o := range orgs {
		out[i] = o.Slug
	}
	return out
}

func warningMessages(warnings []model.Warning) []string {
	out := make([]string, len(warnings))
	for i, w := range warnings {
		out[i] = w.Message
	}
	return out
}

// Membership is the union of both directions, and doing both is a no-op rather than a
// duplicate.
func TestResolveOrganizationsUnionsBothDirections(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [
        {"slug": "tools", "name": "Tools", "description": "Small things", "repos": ["alpha", "beta"], "avatar": "logos/tools.png"},
        {"slug": "apps", "name": "Apps"}
      ]
    }`)
	res := ResolveOrganizations(cfg, []OrgRepoInput{
		{Slug: "alpha", Org: "tools"}, // claimed from both sides
		{Slug: "beta", Org: ""},
		{Slug: "gamma", Org: "apps"},
	})
	if got := orgSlugs(res.Organizations); !reflect.DeepEqual(got, []string{"apps", "tools"}) {
		t.Fatalf("orgs = %v, want them sorted by slug", got)
	}
	apps, tools := res.Organizations[0], res.Organizations[1]
	if !reflect.DeepEqual(apps.Repos, []string{"gamma"}) {
		t.Errorf("apps.repos = %v", apps.Repos)
	}
	if !reflect.DeepEqual(tools.Repos, []string{"alpha", "beta"}) {
		t.Errorf("tools.repos = %v, want sorted and de-duplicated", tools.Repos)
	}
	if tools.Description == nil || *tools.Description != "Small things" {
		t.Errorf("description = %v", tools.Description)
	}
	// The avatar is a public/-relative path, normalised to a leading slash exactly as the
	// TypeScript's PublicPath transform does.
	if tools.Avatar == nil || *tools.Avatar != "/logos/tools.png" {
		t.Errorf("avatar = %v", tools.Avatar)
	}
	if apps.Description != nil || apps.Avatar != nil {
		t.Errorf("an org that declared neither must carry nulls: %+v", apps)
	}
	if len(res.Warnings) != 0 {
		t.Errorf("warnings = %v", warningCodes(res.Warnings))
	}
}

// An org with no members still gets emitted: it has a page either way.
func TestResolveOrganizationsKeepsEmptyOrgs(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [{"slug": "lonely", "name": "Lonely"}]
    }`)
	res := ResolveOrganizations(cfg, nil)
	if len(res.Organizations) != 1 || len(res.Organizations[0].Repos) != 0 {
		t.Fatalf("orgs = %+v", res.Organizations)
	}
	if res.Organizations[0].Repos == nil {
		t.Error("Repos must be an empty slice, not nil — nil marshals as null")
	}
}

// Dangling references on either side are warnings, never failures: a config edit that renames
// a repo must not take the site down.
func TestResolveOrganizationsWarnsAboutDanglingReferences(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [
        {"slug": "zed", "name": "Zed", "repos": ["ghost", "ghost", "alpha"]},
        {"slug": "acme", "name": "Acme", "repos": ["phantom"]}
      ]
    }`)
	res := ResolveOrganizations(cfg, []OrgRepoInput{
		{Slug: "zulu", Org: "nowhere"},
		{Slug: "alpha", Org: "nowhere"},
	})
	// Org-side warnings come in CONFIG order (zed before acme, as written), repo-side ones in
	// SLUG order — so neither depends on how the scans finished.
	if got := warningCodes(res.Warnings); !reflect.DeepEqual(got,
		[]string{"org-unknown-repo", "org-unknown-repo", "repo-unknown-org", "repo-unknown-org"}) {
		t.Fatalf("warnings = %v", got)
	}
	messages := warningMessages(res.Warnings)
	if messages[0] != "organization 'zed' lists repo 'ghost', which was not ingested; ignored" {
		t.Errorf("first warning = %q", messages[0])
	}
	if messages[1] != "organization 'acme' lists repo 'phantom', which was not ingested; ignored" {
		t.Errorf("a slug listed twice must warn once; warnings = %v", messages)
	}
	if messages[2] != "repo 'alpha' declares org 'nowhere', which is not configured; ignored" {
		t.Errorf("repo-side warnings are not in slug order: %v", messages[2:])
	}
	// org-unknown-repo is site-level; repo-unknown-org names the repo that declared it.
	if res.Warnings[0].Repo != nil {
		t.Error("org-unknown-repo must be site-level")
	}
	if res.Warnings[2].Repo == nil || *res.Warnings[2].Repo != "alpha" {
		t.Errorf("repo-unknown-org repo = %v", res.Warnings[2].Repo)
	}
}

// The same slug twice is merged rather than dropped or fatal: the first definition wins and
// the repos lists are unioned.
func TestResolveOrganizationsMergesDuplicateSlugs(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [
        {"slug": "dup", "name": "First", "description": "kept", "repos": ["alpha"]},
        {"slug": "dup", "name": "Second", "description": "dropped", "repos": ["beta"]}
      ]
    }`)
	res := ResolveOrganizations(cfg, []OrgRepoInput{{Slug: "alpha"}, {Slug: "beta"}})
	if len(res.Organizations) != 1 {
		t.Fatalf("orgs = %+v, want one merged entry", res.Organizations)
	}
	org := res.Organizations[0]
	if org.Name != "First" || org.Description == nil || *org.Description != "kept" {
		t.Errorf("the later definition won: %+v", org)
	}
	if !reflect.DeepEqual(org.Repos, []string{"alpha", "beta"}) {
		t.Errorf("repos = %v, want both lists unioned", org.Repos)
	}
}

// A repo's own `org` names one org; `organizations[].repos` is a claim, not an exclusive one.
func TestResolveOrganizationsLetsSeveralOrgsClaimOneRepo(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [
        {"slug": "a", "name": "A", "repos": ["shared"]},
        {"slug": "b", "name": "B", "repos": ["shared"]}
      ]
    }`)
	res := ResolveOrganizations(cfg, []OrgRepoInput{{Slug: "shared"}})
	for _, org := range res.Organizations {
		if !reflect.DeepEqual(org.Repos, []string{"shared"}) {
			t.Errorf("%s.repos = %v", org.Slug, org.Repos)
		}
	}
}

// The result must not depend on the order the caller happened to hand repos over.
func TestResolveOrganizationsIgnoresRepoInputOrder(t *testing.T) {
	cfg := testResolvedConfig(t, t.TempDir(), `{
      "owner": {"name": "Tester", "handle": "tester"},
      "organizations": [{"slug": "tools", "name": "Tools"}]
    }`)
	forwards := ResolveOrganizations(cfg, []OrgRepoInput{
		{Slug: "alpha", Org: "tools"}, {Slug: "beta", Org: "gone"}, {Slug: "aardvark", Org: "gone"},
	})
	backwards := ResolveOrganizations(cfg, []OrgRepoInput{
		{Slug: "aardvark", Org: "gone"}, {Slug: "beta", Org: "gone"}, {Slug: "alpha", Org: "tools"},
	})
	if !reflect.DeepEqual(forwards, backwards) {
		t.Fatalf("input order changed the result:\n%+v\n%+v", forwards, backwards)
	}
	if got := warningMessages(forwards.Warnings); got[0] != "repo 'aardvark' declares org 'gone', which is not configured; ignored" {
		t.Errorf("warnings = %v, want slug order", got)
	}
}
